ree/reeEstimateError.cc

```diff
@@ -0,0 +1,458 @@
+      double weight = apf::getIntWeight(fme, 2*order-1, n);
+      apf::Matrix3x3 fJ;
+      apf::getJacobian(fme, p, fJ);
+      double jdet = apf::getJacobianDeterminant(
+          fJ, apf::getDimension(mesh, face));
+
+      // obtain corrected flux vector
+      double comp[nc];
+      apf::getComponents(flux_field, e, n, comp);
+      apf::Vector3 theta_plus_tk;
+      int index = ii*3;
+      theta_plus_tk[0] = comp[index];
+      theta_plus_tk[1] = comp[index+1];
+      theta_plus_tk[2] = comp[index+2];
+
+      // compute p+1 order 3D vector shapes
+      apf::NewArray<apf::Vector3> tetVectorShapes (nedofs);
+      apf::MeshElement* me = apf::createMeshElement(mesh, e);
+      apf::Element* el = apf::createElement(fp1, me);
+      apf::Vector3 tetxi = apf::boundaryToElementXi(mesh, face, e, p);
+      apf::getVectorShapeValues(el, tetxi, tetVectorShapes);
+
+      apf::destroyElement(el);
+      apf::destroyMeshElement(me);
+
+      // compute integral
+      double w = weight * jdet;
+      theta_plus_tk = theta_plus_tk * w;
+
+      // matrix vector multiplication
+      for (int i = 0; i < nedofs; i++)
+        lambda(i) += theta_plus_tk * tetVectorShapes[i];
+
+    } // end integral loop
+    apf::destroyMeshElement(fme);
+  } // end face loop
+}
+
+static void getEssentialElementNDDofs(apf::Mesh* mesh, apf::MeshEntity* e,
+    apf::Field* f, apf::NewArray<int>& ess_dofs, apf::NewArray<int>& uness_dofs)
+{
+  apf::FieldShape* fs = f->getShape();
+  int etype = mesh->getType(e);
+  PCU_ALWAYS_ASSERT(etype == apf::Mesh::TET);
+  int nedofs = apf::countElementNodes(fs, etype);
+
+  apf::NewArray<int> marker_list (nedofs);
+  for (int i = 0; i < nedofs; i++)
+    marker_list[i] = 0;
+
+  // populate marker list by iterating over downward edges and faces
+  apf::Downward edges;
+  int nedges = mesh->getDownward(e, 1, edges);
+  PCU_ALWAYS_ASSERT(nedges == 6);
+  for (int i = 0; i < nedges; i++) {
+    int nodesOnEdge = fs->countNodesOn(mesh->getType(edges[i]));
+    if ( isOnDomainBoundary(mesh, edges[i]) ) {
+      for (int n = 0; n < nodesOnEdge; n++) {
+        marker_list[(i*nodesOnEdge)+n] = -1;
+      }
+    }
+  }
+  int nodesOnEdges = fs->countNodesOn(apf::Mesh::EDGE) * nedges;
+
+  apf::Downward faces;
+  int nfaces = mesh->getDownward(e, 2, faces);
+  PCU_ALWAYS_ASSERT(nfaces == 4);
+  for (int i = 0; i < nfaces; i++) {
+    int nodesOnFace = fs->countNodesOn(mesh->getType(faces[i]));
+    if ( isOnDomainBoundary(mesh, faces[i]) ) {
+      for (int n = 0; n < nodesOnFace; n++) {
+        marker_list[(nodesOnEdges + i*nodesOnFace)+n] = -1;
+      }
+    }
+  }
+
+  int num_marked = 0;
+  for (int i = 0; i < nedofs; i++) {
+    if (marker_list[i])
+      num_marked++;
+  }
+
+  // use marker list to get ess_dofs list
+  ess_dofs.allocated() ? ess_dofs.resize(num_marked)
+    : ess_dofs.allocate(num_marked);
+  uness_dofs.allocated() ? uness_dofs.resize(nedofs-num_marked)
+    : uness_dofs.allocate(nedofs-num_marked);
```

Comment:
I just called `resize(...)` on an unallocated `apf::NewArray` and it seems to work. While writing this code, I had followed the convention I saw being used in `core` so that is why I have the `if` conditional. I can remove it if that's better?

Reply:
this is the code for `resize`

```
    void resize(unsigned n)
    {   
      if (sz == n)
        return;
      delete [] elems;
      sz = n;
      elems = new T[n];
    }                    
```

My understanding is that calling `delete [] elems;' on `elems` without having called `new` could result in undefined behavior.

Reply:
@mortezah  So, is it better to check if the `apf::NewArray' is allocated before calling resize on it?

Reply:
If the pointer is `NULL` then calling `delete` is safe:
https://stackoverflow.com/questions/13818803/check-for-null-before-delete-in-c-good-practice

`NewArray` is derived from the run-time dynamic `Array` in `can/canArray.h` and the constructor either sets the pointer to `0` (equivalent to `NULL` as far as `delete` is concerned):

https://github.com/SCOREC/core/blob/b31683d228d59d98bd21cfb978ce96cf1419a997/can/canArray.h#L60-L62

or allocates the `elems` to the given size.  I'd suggest just calling `resize(...)`.

Reply:
Okay. Thanks for the explanation. I have removed the allocation calls and just used `resize(...)`.